presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -184,6 +180,73 @@ else if (type instanceof TimestampWithTimeZoneType) {
         return Optional.of(unmodifiableList(row));
     }
 
+    private Object roundParametricDateTime(Object value, Type type)
+    {
+        if (supportsParametricDateTime || value == null) {
+            return value;
+        }
+
+        if (type instanceof TimestampType) {
+            return ((SqlTimestamp) value).roundTo(3);
+        }
+
+        if (type instanceof TimestampWithTimeZoneType) {
+            return ((SqlTimestampWithTimeZone) value).roundTo(3);
+        }
+
+        if (type instanceof ArrayType) {
+            List<Object> values = (List<Object>) value;
+
+            ImmutableList.Builder<Object> result = ImmutableList.builder();
+            Type elementType = ((ArrayType) type).getElementType();
+
+            for (int i = 0; i < values.size(); i++) {
+                result.add(roundParametricDateTime(values.get(i), elementType));
```

Comment:
we shouldn't need to copy the array data, if it does not contain a timestamp type

you can effect that pretty concisely by making the method return optional
where empty would signal "no change"

this way you would do the decomposition only once

Reply:
it can't be static, I'd need to pass supportsParametricDateTime everywhere